hera/character/Portrait.tsx

```diff
@@ -30,43 +29,64 @@ export default memo(function Portrait({
   unit: UnitInfo;
   variant?: number;
 }) {
+  const ref = useRef<HTMLCanvasElement>(null);
   const hasPortraits = useSprites('portraits');
-  const sprite = hasPortraits
-    ? spriteURL('Portraits', encodeDynamicPlayerID(player))
-    : '';
 
   const { position } = unit.sprite.portrait;
-  const y = -(position.y + (variant || 0)) * PortraitHeight + 'px';
-  const alternateY = -(position.y + (variant || 0) + 6) * PortraitHeight + 'px';
+  const keyFrameYs = [
+    (position.y + (variant || 0)) * PortraitHeight,
+    (position.y + (variant || 0) + 6) * PortraitHeight,
+  ];
+  let curKeyFrameY = 0;
+
+  useLayoutEffect(() => {
+    if (!hasPortraits) {
+      return;
+    }
+
+    const canvas = ref.current;
+    if (!canvas) {
+      return;
+    }
+
+    const image = spriteImage('Portraits', encodeDynamicPlayerID(player));
+    const context = canvas.getContext('2d')!;
+
+    function animate() {
+      context.drawImage(
+        image,
+        position.x * PortraitWidth,
+        keyFrameYs[curKeyFrameY],
+        PortraitWidth,
+        PortraitHeight,
+        0,
+        0,
+        PortraitWidth,
+        PortraitHeight,
+      );
+
+      curKeyFrameY = (curKeyFrameY + 1) % keyFrameYs.length;
+    }
+
+    if (animate) {
+      setInterval(animate, 1000 / keyFrameYs.length);
+    }
+  }, [hasPortraits]);
 
   return (
     <div
       className={cx(portraitStyle, clip && clipStyle)}
       style={{
-        [vars.set('x')]: -position.x * PortraitWidth + 'px',
-        [vars.set('y')]: y,
-        [vars.set('alternate-y')]: alternateY,
         height: PortraitHeight,
         width: PortraitWidth,
         zoom: scale,
       }}
     >
-      {sprite && (
-        <img
-          className={cx(
-            spriteStyle,
-            animate && animateStyle,
-            paused && pausedStyle,
-          )}
-          src={sprite}
-        />
-      )}
+      <canvas ref={ref}></canvas>
     </div>
   );
 });
 
-const vars = new CSSVariables<'x' | 'y' | 'alternate-y'>('p');
-
 const portraitStyle = css`
   contain: content;
```

Comment:
I think we don't need this line since we aren't overflowing the element anymore.

Reply:
[feat: update portraitStyle](https://github.com/nkzw-tech/athena-crisis/pull/32/commits/e2e993f81cf79752bd117825d9730997e7b51321)